java/statestore/src/main/java/sleeper/statestore/dynamodb/DynamoDBFileReferenceStore.java

```diff
@@ -293,6 +365,22 @@ public void atomicallyAssignJobIdToFileReferences(String jobId, List<FileReferen
             double totalConsumed = consumedCapacity.stream().mapToDouble(ConsumedCapacity::getCapacityUnits).sum();
             LOGGER.debug("Updated job status of {} files, read capacity consumed = {}",
                     files.size(), totalConsumed);
+        } catch (TransactionCanceledException e) {
+            List<CancellationReason> reasons = e.getCancellationReasons();
+            for (int i = 0; i < files.size(); i++) {
+                CancellationReason reason = reasons.get(i);
+                if (isConditionCheckFailure(reason)) {
+                    FileReference fileReference = files.get(i);
+                    if (reason.getItem() != null) {
+                        FileReference failedUpdate = fileReferenceFormat.getFileReferenceFromAttributeValues(reason.getItem());
+                        if (failedUpdate.getJobId() != null) {
+                            throw new FileReferenceAssignedToJobException(failedUpdate);
+                        }
+                    }
+                    throw new FileReferenceNotFoundException(fileReference);
+                }
+            }
```

Comment:
This can be its own method to process the cancellation reasons, or split further. We can avoid discarding the original TransactionCanceledException.

Reply:
This also ignores the failure entirely if it's not a condition check failure. That case needs to be rethrown as a StateStoreException.